pyfar/dsp/filter/gammatone.py

```diff
@@ -0,0 +1,581 @@
+            The imaginary part of the complex output signal. This approximates
+            the Hilbert transform of the band-limited Gammatone filter output.
+
+        Notes
+        -----
+        -``sqrt(real.time**2 + imag.time**2)`` gives the envelope of the
+          Gammatone filter output.
+        - If the cshape of the output signals ``real.cshape`` and
+          ``imag.cshape`` generally is ``(self.n_bands, ) + signal.cshape``.
+        - An exception to this occurs if ``signal.cshape`` is ``(1, )``, i.e.,
+          signal is a single channel signal. In this case the cshape of the
+          output signals is ``(self.n_bands)`` and `not` ``(self.n_bands, 1)``.
+        """
+
+        # check input
+        if not isinstance(signal, pf.Signal):
+            raise TypeError("signal must be a pyfar Signal object")
+        if signal.sampling_rate != self.sampling_rate:
+            raise ValueError(("The sampling rates of the signal and Gammatone"
+                              " filter bank do not match"))
+
+        # prepare multi-dimensional signals
+        time_in = signal.flatten().time
+        time_out = np.zeros((self.n_bands, ) + time_in.shape, dtype=complex)
+
+        # reset or initialize the state as a list of as many zero arrays as
+        # the filter bank has bands
+        if reset_state or self._state is None:
+            state = np.zeros((4, time_in.shape[0], 2), dtype=complex)
+            self._state = [state for _ in range(self.n_bands)]
+        elif len(self._state) != self.n_bands \
+                or self._state[0].shape != (4, time_in.shape[0], 2):
+            raise ValueError((
+                "The shape of the signal and the internal state of the filter "
+                "bank do not match. Try calling process with reset_state=True "
+                "or with the signal that it was previously used with."
+            ))
+
+        # apply the filter band by band
+        # using second order sections is faster than a manual call of
+        # sgn.lfilter four time in a row
+        for bb in range(self.n_bands):
+
+            sos_section = np.tile(np.atleast_2d(
+                [1, 0, 0, 1, -self._coefficients[bb], 0]),
+                (4, 1))
+            sos_section[3, 0] = self._normalizations[bb]
+            time_out[bb], self._state[bb] = sgn.sosfilt(
+                sos_section, time_in, axis=-1, zi=self._state[bb])
+
+        # restore original channel shape
+        time_out = np.reshape(time_out,
+                              (self.n_bands, ) + signal.cshape + (-1, ))
+
+        # squeeze dimension of single channel signal
+        if signal.cshape == (1, ):
+            time_out = np.squeeze(time_out)
+
+        # return real and immaginary part of output as pyfar Signal objects
+        real = signal.copy()
+        real.time = np.real(time_out)
+        imag = signal.copy()
+        imag.time = np.imag(time_out)
+
+        return real, imag
+
+    def sum(self, real, imag):
```

Comment:
Sum is very misleading in this context, because it's a very specific sum. This is the reconstruction step, isn't it?